src/shave.cpp

```diff
@@ -350,7 +351,26 @@ void filterFeatures(vtzero::tile_builder* finalvt,
                 feature_builder.set_id(feature.id());
             }
             feature_builder.set_geometry(feature.geometry());
+
+            std::string property_filter_type = property_filter.type;
+            std::vector<std::string> properties = property_filter.values;
+            auto keytable = layer.key_table();
+
+            bool needAllProperties = property_filter_type == "all";
+            // if (property_filter_type != "all") {
             while (auto idxs = feature.next_property_indexes()) {
+                if (!needAllProperties) {
+                    // get the key only if we don't need all the properties;
+                    std::string key = keytable[idxs.key().value()].to_string();
```

Comment:
How about - for best performance - instead of comparing `std::string` here, we compare `uint32_t`? I think this could be accomplished by reversing the `properties` before the `next_property_indexes()` loop. You'd create a `property_idxs` vector that represents the indexs for all the property names. Then inside the `next_property_indexes()` you could see if the index value is in that `property_idxs` vector.

Reply:
following the topic we talked in the channel,I use the new method to do this job:

```
auto const& keytable = layer.key_table();
std::vector<std::ptrdiff_t> props_by_index;
for (auto const& prop : properties) {
   auto itr = std::find(keytable.begin(), keytable.end(), prop);
   if (itr != keytable.end()) {
        props_by_index.emplace_back(std::distance(keytable.begin(), itr));
    }
}
```
and
```
if (std::find(props_by_index.begin(), props_by_index.end(), idxs.key()) == props_by_index.end()) {
    continue;
}
```

which can save about **5%** resource.